Missing ts-loader breaks npm run dll:build

Hello there,

this generator breaks as soon as I tried to generate a ts package.

I also had to `export NODE_OPTIONS=--openssl-legacy-provider` as suggested [here](https://stackoverflow.com/a/69699772/1679665) to make it start, which surprised me [given this closed issue](https://github.com/ckeditor/ckeditor5-package-generator/pull/138), but then it broke because `ts-loader` is missing. I noticed that it is in the generated lock file but not in the generate package.json

After `npm i -D ts-loader` the `npm run dll:build` works.

## 📝 Provide detailed reproduction steps (if any)

```bash
npx ckeditor5-package-generator @mytest/ckeditor5-a-package --use-npm --lang ts --verbose
```

### ✔️ Expected result

Webpack compiles correctly

### ❌ Actual result

```bash
npx ckeditor5-package-generator @mytest/ckeditor5-a-package --use-npm --lang ts --verbose
📍 Verifying the specified package name.
📍 Checking whether the "ckeditor5-a-package" directory can be created.
📍 Creating the directory "/Users/mac-user/Development/ckeditor/ckeditor5-a-package".
📍 Collecting the latest CKEditor 5 packages versions...
📍 Copying files...
* Copying "common/.editorconfig"...
* Copying "common/.stylelintrc"...
* Copying "common/ckeditor5-metadata.json"...
* Copying "common/lang/contexts.json"...
* Copying "common/LICENSE.md"...
* Copying "common/README.md"...
* Copying "common/theme/icons/ckeditor.svg"...
* Copying "ts/.eslintrc.js"...
* Copying "ts/.gitattributes"...
* Copying "ts/.gitignore.txt"...
* Copying "ts/package.json"...
* Copying "ts/sample/ckeditor.ts.txt"...
* Copying "ts/sample/dll.html.txt"...
* Copying "ts/sample/index.html"...
* Copying "ts/src/_PLACEHOLDER_.ts.txt"...
* Copying "ts/src/index.ts.txt"...
* Copying "ts/tests/_PLACEHOLDER_.ts.txt"...
* Copying "ts/tests/index.ts.txt"...
* Copying "ts/tsconfig.json.txt"...
* Copying "ts/tsconfig.release.json"...
* Copying "ts/typings/ckeditor5-inspector.d.ts"...
* Copying "ts/typings/types.d.ts.txt"...
📍 Installing dependencies... It takes a while.
npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated @stylelint/postcss-markdown@0.36.2: Use the original unforked package instead: postcss-markdown
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated debug@3.2.6: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated @stylelint/postcss-css-in-js@0.37.3: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.

> @mytest/ckeditor5-a-package@0.0.1 prepare
> npm run dll:build


> @mytest/ckeditor5-a-package@0.0.1 dll:build
> ckeditor5-package-tools dll:build

(node:58963) [DEP_WEBPACK_MAIN_TEMPLATE_HASH_FOR_CHUNK] DeprecationWarning: MainTemplate.hooks.hashForChunk is deprecated (use JavascriptModulesPlugin.getCompilationHooks().chunkHash instead)
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:58963) [DEP_WEBPACK_CHUNK_TEMPLATE_HASH_FOR_CHUNK] DeprecationWarning: ChunkTemplate.hooks.hashForChunk is deprecated (use JavascriptModulesPlugin.getCompilationHooks().chunkHash instead)
(node:58963) [DEP_WEBPACK_COMPILATION_OPTIMIZE_CHUNK_ASSETS] DeprecationWarning: optimizeChunkAssets is deprecated (use Compilation.hooks.processAssets instead and use one of Compilation.PROCESS_ASSETS_STAGE_* as stage option)
/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/lib/tasks/dll-build.js:27
				reject( new Error( stats.toString() ) );
				        ^

Error: assets by status 0 bytes [cached] 1 asset

ERROR in main
Module not found: Error: Can't resolve 'ts-loader' in '/Users/mac-user/Development/ckeditor/ckeditor5-a-package'
resolve 'ts-loader' in '/Users/mac-user/Development/ckeditor/ckeditor5-a-package'
  Parsed request is a module
  using description file: /Users/mac-user/Development/ckeditor/ckeditor5-a-package/package.json (relative path: .)
    resolve as module
      looking for modules in /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules
        single file module
          using description file: /Users/mac-user/Development/ckeditor/ckeditor5-a-package/package.json (relative path: ./node_modules/ts-loader)
            no extension
              /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/ts-loader doesn't exist
            .js
              /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/ts-loader.js doesn't exist
        /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/ts-loader doesn't exist
      /Users/mac-user/Development/ckeditor/node_modules doesn't exist or is not a directory
      /Users/mac-user/Development/node_modules doesn't exist or is not a directory
      /Users/mac-user/Development/Websites/next-cms-context/node_modules doesn't exist or is not a directory
      /Users/mac-user/Development/Websites/node_modules doesn't exist or is not a directory
      /Users/mac-user/Development/node_modules doesn't exist or is not a directory
      /Users/mac-user/node_modules doesn't exist or is not a directory
      /Users/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory

webpack 5.77.0 compiled with 1 error in 150 ms
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/lib/tasks/dll-build.js:27:13
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/webpack.js:148:8
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/HookWebpackError.js:68:3
    at Hook.eval [as callAsync] (eval at create (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/Hook.js:18:14)
    at Cache.shutdown (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Cache.js:150:23)
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Compiler.js:1225:15
    at Hook.eval [as callAsync] (eval at create (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/Hook.js:18:14)
    at Compiler.close (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Compiler.js:1218:23)
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/webpack.js:147:16
    at finalCallback (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Compiler.js:441:32)
    at /Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Compiler.js:458:13
    at Hook.eval [as callAsync] (eval at create (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/tapable/lib/Hook.js:18:14)
    at onCompiled (/Users/mac-user/Development/ckeditor/ckeditor5-a-package/node_modules/@ckeditor/ckeditor5-package-tools/node_modules/webpack/lib/Compiler.js:456:21)

Node.js v18.12.1
npm ERR! code 1
npm ERR! path /Users/mac-user/Development/ckeditor/ckeditor5-a-package
npm ERR! command failed
npm ERR! command sh -c npm run dll:build

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mac-user/.npm/_logs/2023-03-30T13_26_57_129Z-debug-0.log
/Users/mac-user/.npm/_npx/5166aebc640c7801/node_modules/ckeditor5-package-generator/lib/utils/install-dependencies.js:77
				return reject( new Error( 'Installing dependencies finished with an error.' ) );
				               ^

Error: Installing dependencies finished with an error.
    at ChildProcess.<anonymous> (/Users/mac-user/.npm/_npx/5166aebc640c7801/node_modules/ckeditor5-package-generator/lib/utils/install-dependencies.js:77:20)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)

Node.js v18.12.1
```

### ❓ Possible solution

* include ts-loader in the dev dependencies of the generated process if needed by webpack 
* solve the annoying issue of "legacy provider" by prepending NODE_OPTIONS=--openssl-legacy-provider by default or make node 18 (current LTS) as default

## 📃 Other details

* Node: 18.12.1
* Npm: 9.5.1
* First affected version: v1.0.0-beta.10 (didn't try older)

macOS 13.2.1 on M1 Pro
---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

--- a/packages/ckeditor5-package-generator/lib/utils/get-package-name-formats.js
+++ b/packages/ckeditor5-package-generator/lib/utils/get-package-name-formats.js
@@ -90,13 +90,13 @@ function uppercaseFirstChar( string ) {
 /**
  * @typedef {Object} FormattedName
  *
- * @property {String} raw: super-feature-name
+ * @property {String} raw super-feature-name
  *
- * @property {String} spacedOut Super plugin name
+ * @property {String} spacedOut Super feature name
  *
- * @property {String} camelCase superPluginName
+ * @property {String} camelCase superFeatureName
  *
- * @property {String} pascalCase SuperPluginName
+ * @property {String} pascalCase SuperFeatureName
  *
- * @property {String} lowerCaseMerged superpluginname
+ * @property {String} lowerCaseMerged superfeaturename
  */

--- a/packages/ckeditor5-package-tools/lib/utils/get-karma-config.js
+++ b/packages/ckeditor5-package-tools/lib/utils/get-karma-config.js
@@ -8,7 +8,9 @@
 /* eslint-env node */
 
 const path = require( 'path' );
-const { loaderDefinitions } = require( './webpack-utils' );
+const { loaderDefinitions, getModuleResolutionPaths } = require( './webpack-utils' );
+
+const PACKAGE_ROOT_DIR = path.join( __dirname, '..', '..' );
 
 /**
  * @param {Object} options
@@ -139,12 +141,15 @@ module.exports = options => {
  * @returns {Object}
  */
 function getWebpackConfiguration( options ) {
+	const moduleResolutionPaths = getModuleResolutionPaths( PACKAGE_ROOT_DIR );
+
 	const config = {
 		mode: 'development',
 
 		resolve: {
 			// Triple dots syntax allows extending default extension list instead of overwriting it.
-			extensions: [ '.ts', '...' ]
+			extensions: [ '.ts', '...' ],
+			modules: moduleResolutionPaths
 		},
 
 		module: {
@@ -156,9 +161,7 @@ function getWebpackConfiguration( options ) {
 		},
 
 		resolveLoader: {
-			modules: [
-				'node_modules'
-			]
+			modules: moduleResolutionPaths
 		}
 	};
 

--- a/packages/ckeditor5-package-tools/lib/utils/get-webpack-config-dll.js
+++ b/packages/ckeditor5-package-tools/lib/utils/get-webpack-config-dll.js
@@ -13,7 +13,9 @@ const fs = require( 'fs' );
 const webpack = require( 'webpack' );
 const TerserPlugin = require( 'terser-webpack-plugin' );
 const { CKEditorTranslationsPlugin } = require( '@ckeditor/ckeditor5-dev-translations' );
-const { loaderDefinitions } = require( './webpack-utils' );
+const { loaderDefinitions, getModuleResolutionPaths } = require( './webpack-utils' );
+
+const PACKAGE_ROOT_DIR = path.join( __dirname, '..', '..' );
 
 module.exports = options => {
 	const packageJson = require( path.join( options.cwd, 'package.json' ) );
@@ -58,6 +60,8 @@ module.exports = options => {
 	const entryFileName = fs.readdirSync( path.join( options.cwd, 'src' ) )
 		.find( filePath => /^index\.[jt]s$/.test( filePath ) );
 
+	const moduleResolutionPaths = getModuleResolutionPaths( PACKAGE_ROOT_DIR );
+
 	return {
 		mode: 'production',
 
@@ -90,7 +94,12 @@ module.exports = options => {
 
 		resolve: {
 			// Triple dots syntax allows extending default extension list instead of overwriting it.
-			extensions: [ '.ts', '...' ]
+			extensions: [ '.ts', '...' ],
+			modules: moduleResolutionPaths
+		},
+
+		resolveLoader: {
+			modules: moduleResolutionPaths
 		},
 
 		module: {

--- a/packages/ckeditor5-package-tools/lib/utils/get-webpack-config-server.js
+++ b/packages/ckeditor5-package-tools/lib/utils/get-webpack-config-server.js
@@ -11,7 +11,9 @@ const fs = require( 'fs' );
 const path = require( 'path' );
 const webpack = require( 'webpack' );
 const { CKEditorTranslationsPlugin } = require( '@ckeditor/ckeditor5-dev-translations' );
-const { loaderDefinitions } = require( './webpack-utils' );
+const { loaderDefinitions, getModuleResolutionPaths } = require( './webpack-utils' );
+
+const PACKAGE_ROOT_DIR = path.join( __dirname, '..', '..' );
 
 module.exports = options => {
 	const webpackPlugins = [
@@ -35,6 +37,8 @@ module.exports = options => {
 
 	const entryFileName = fs.readdirSync( path.join( options.cwd, 'sample' ) )
 		.find( filePath => /^ckeditor\.[jt]s$/.test( filePath ) );
+
+	const moduleResolutionPaths = getModuleResolutionPaths( PACKAGE_ROOT_DIR );
 
 	return {
 		mode: options.production ? 'production' : 'development',
@@ -65,7 +69,12 @@ module.exports = options => {
 
 		resolve: {
 			// Triple dots syntax allows extending default extension list instead of overwriting it.
-			extensions: [ '.ts', '...' ]
+			extensions: [ '.ts', '...' ],
+			modules: moduleResolutionPaths
+		},
+
+		resolveLoader: {
+			modules: moduleResolutionPaths
 		},
 
 		module: {

--- a/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
@@ -64,5 +64,16 @@ module.exports = {
 				]
 			};
 		}
+	},
+	getModuleResolutionPaths: packageRootDir => {
+		return [
+			'node_modules',
+			//  Root of this package.
+			path.resolve( packageRootDir, 'node_modules' ),
+			// Root of the CPG repository.
+			path.resolve( packageRootDir, '..', '..', 'node_modules' ),
+			// Root of the repository that uses this package.
+			path.resolve( packageRootDir, '..', '..', '..', 'node_modules' )
+		];
 	}
 };

--- a/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
@@ -68,8 +68,10 @@ module.exports = {
 	getModuleResolutionPaths: packageRootDir => {
 		return [
 			'node_modules',
-			path.resolve( packageRootDir, '..', '..', 'node_modules' ),
-			path.resolve( packageRootDir, '..', '..', '..', 'node_modules' )
+			//  Root of this package.
+			path.resolve( packageRootDir, 'node_modules' ),
+			// Root of the CPG repository.
+			path.resolve( packageRootDir, '..', '..', 'node_modules' )
 		];
 	}
 };

--- a/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
@@ -68,7 +68,8 @@ module.exports = {
 	getModuleResolutionPaths: packageRootDir => {
 		return [
 			'node_modules',
-			path.resolve( packageRootDir, '..', '..', 'node_modules' )
+			path.resolve( packageRootDir, '..', '..', 'node_modules' ),
+			path.resolve( packageRootDir, '..', '..', '..', 'node_modules' )
 		];
 	}
 };

--- a/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
@@ -68,12 +68,7 @@ module.exports = {
 	getModuleResolutionPaths: packageRootDir => {
 		return [
 			'node_modules',
-			//  Root of this package.
-			path.resolve( packageRootDir, 'node_modules' ),
-			// Root of the CPG repository.
-			path.resolve( packageRootDir, '..', '..', 'node_modules' ),
-			// Root of the repository that uses this package.
-			path.resolve( packageRootDir, '..', '..', '..', 'node_modules' )
+			path.resolve( packageRootDir, '..', '..', 'node_modules' )
 		];
 	}
 };

--- a/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/lib/utils/webpack-utils.js
@@ -68,10 +68,7 @@ module.exports = {
 	getModuleResolutionPaths: packageRootDir => {
 		return [
 			'node_modules',
-			//  Root of this package.
-			path.resolve( packageRootDir, 'node_modules' ),
-			// Root of the CPG repository.
-			path.resolve( packageRootDir, '..', '..', 'node_modules' )
+			path.resolve( packageRootDir, 'node_modules' )
 		];
 	}
 };

--- a/packages/ckeditor5-package-tools/package.json
+++ b/packages/ckeditor5-package-tools/package.json
@@ -40,6 +40,7 @@
     "stylelint": "^13.13.1",
     "stylelint-config-ckeditor5": "^2.0.1",
     "terser-webpack-plugin": "^3.0.2",
+    "ts-loader": "^9.4.2",
     "ts-node": "^10.9.1",
     "typescript": "^4.7.4",
     "webpack": "^5.58.1",

--- a/packages/ckeditor5-package-tools/tests/utils/get-karma-config.js
+++ b/packages/ckeditor5-package-tools/tests/utils/get-karma-config.js
@@ -8,6 +8,7 @@
 const sinon = require( 'sinon' );
 const mockery = require( 'mockery' );
 const expect = require( 'chai' ).expect;
+const path = require( 'path' );
 
 describe( 'lib/utils/get-karma-config', () => {
 	let getKarmaConfig, stubs;
@@ -35,7 +36,8 @@ describe( 'lib/utils/get-karma-config', () => {
 					styles: sinon.stub(),
 					coverage: sinon.stub(),
 					typescript: sinon.stub()
-				}
+				},
+				getModuleResolutionPaths: sinon.stub()
 			}
 		};
 
@@ -43,6 +45,7 @@ describe( 'lib/utils/get-karma-config', () => {
 		stubs.webpackUtils.loaderDefinitions.typescript.returns( 'typescript-loader' );
 		stubs.webpackUtils.loaderDefinitions.styles.withArgs( cwd ).returns( 'styles-loader' );
 		stubs.webpackUtils.loaderDefinitions.coverage.withArgs( cwd ).returns( 'coverage-loader' );
+		stubs.webpackUtils.getModuleResolutionPaths.returns( 'loader-resolution-paths' );
 
 		mockery.registerMock( 'path', stubs.path );
 		mockery.registerMock( './webpack-utils', stubs.webpackUtils );
@@ -196,6 +199,12 @@ describe( 'lib/utils/get-karma-config', () => {
 	} );
 
 	describe( 'webpack configuration', () => {
+		it( 'has mode set to "development"', () => {
+			const config = getKarmaConfig( { cwd } );
+
+			expect( config.webpack.mode ).to.equal( 'development' );
+		} );
+
 		it( 'uses correct loaders', () => {
 			const config = getKarmaConfig( { cwd } );
 
@@ -223,16 +232,25 @@ describe( 'lib/utils/get-karma-config', () => {
 			expect( config.webpack.resolve.extensions ).to.deep.equal( [ '.ts', '...' ] );
 		} );
 
-		it( 'has mode set to "development"', () => {
+		it( 'resolves correct module paths', () => {
 			const config = getKarmaConfig( { cwd } );
 
-			expect( config.webpack.mode ).to.equal( 'development' );
+			expect( config.webpack.resolve ).to.have.property( 'modules', 'loader-resolution-paths' );
 		} );
 
-		it( 'has node_modules defined in "resolveLoader"', () => {
+		it( 'resolves correct loader paths', () => {
 			const config = getKarmaConfig( { cwd } );
 
-			expect( config.webpack.resolveLoader ).to.deep.equal( { modules: [ 'node_modules' ] } );
+			expect( config.webpack.resolveLoader ).to.have.property( 'modules', 'loader-resolution-paths' );
+		} );
+
+		it( 'calls "getModuleResolutionPaths" with correct arguments', () => {
+			getKarmaConfig( { cwd } );
+
+			expect( stubs.webpackUtils.getModuleResolutionPaths.callCount ).to.equal( 1 );
+			const normalizedArgument = stubs.webpackUtils.getModuleResolutionPaths.firstCall.firstArg
+				.split( path.sep ).join( path.posix.sep );
+			expect( normalizedArgument.endsWith( '/packages/ckeditor5-package-tools/lib/utils/../..' ) ).to.equal( true );
 		} );
 
 		it( 'allows enabling source maps', () => {

--- a/packages/ckeditor5-package-tools/tests/utils/get-webpack-config-dll.js
+++ b/packages/ckeditor5-package-tools/tests/utils/get-webpack-config-dll.js
@@ -8,6 +8,7 @@
 const mockery = require( 'mockery' );
 const sinon = require( 'sinon' );
 const expect = require( 'chai' ).expect;
+const path = require( 'path' );
 
 describe( 'lib/utils/get-webpack-config-dll', () => {
 	let getWebpackConfigDll, stubs;
@@ -47,13 +48,15 @@ describe( 'lib/utils/get-webpack-config-dll', () => {
 					raw: sinon.stub(),
 					styles: sinon.stub(),
 					typescript: sinon.stub()
-				}
+				},
+				getModuleResolutionPaths: sinon.stub()
 			}
 		};
 
 		stubs.webpackUtils.loaderDefinitions.raw.returns( 'raw-loader' );
 		stubs.webpackUtils.loaderDefinitions.typescript.returns( 'typescript-loader' );
 		stubs.webpackUtils.loaderDefinitions.styles.withArgs( cwd ).returns( 'styles-loader' );
+		stubs.webpackUtils.getModuleResolutionPaths.returns( 'loader-resolution-paths' );
 
 		stubs.webpack.DllReferencePlugin = function( ...args ) {
 			return stubs.dllReferencePlugin( ...args );
@@ -103,6 +106,27 @@ describe( 'lib/utils/get-webpack-config-dll', () => {
 		const config = getWebpackConfigDll( { cwd } );
 
 		expect( config.resolve.extensions ).to.deep.equal( [ '.ts', '...' ] );
+	} );
+
+	it( 'resolves correct module paths', () => {
+		const config = getWebpackConfigDll( { cwd } );
+
+		expect( config.resolve ).to.have.property( 'modules', 'loader-resolution-paths' );
+	} );
+
+	it( 'resolves correct loader paths', () => {
+		const config = getWebpackConfigDll( { cwd } );
+
+		expect( config.resolveLoader ).to.have.property( 'modules', 'loader-resolution-paths' );
+	} );
+
+	it( 'calls "getModuleResolutionPaths" with correct arguments', () => {
+		getWebpackConfigDll( { cwd } );
+
+		expect( stubs.webpackUtils.getModuleResolutionPaths.callCount ).to.equal( 1 );
+		const normalizedArgument = stubs.webpackUtils.getModuleResolutionPaths.firstCall.firstArg
+			.split( path.sep ).join( path.posix.sep );
+		expect( normalizedArgument.endsWith( '/packages/ckeditor5-package-tools/lib/utils/../..' ) ).to.equal( true );
 	} );
 
 	it( 'processes the "index.js" file', () => {

--- a/packages/ckeditor5-package-tools/tests/utils/get-webpack-config-server.js
+++ b/packages/ckeditor5-package-tools/tests/utils/get-webpack-config-server.js
@@ -8,6 +8,7 @@
 const mockery = require( 'mockery' );
 const sinon = require( 'sinon' );
 const expect = require( 'chai' ).expect;
+const path = require( 'path' );
 
 describe( 'lib/utils/get-webpack-config-server', () => {
 	let getWebpackConfigServer, stubs;
@@ -43,13 +44,15 @@ describe( 'lib/utils/get-webpack-config-server', () => {
 					raw: sinon.stub(),
 					styles: sinon.stub(),
 					typescript: sinon.stub()
-				}
+				},
+				getModuleResolutionPaths: sinon.stub()
 			}
 		};
 
 		stubs.webpackUtils.loaderDefinitions.raw.returns( 'raw-loader' );
 		stubs.webpackUtils.loaderDefinitions.typescript.returns( 'typescript-loader' );
 		stubs.webpackUtils.loaderDefinitions.styles.withArgs( cwd ).returns( 'styles-loader' );
+		stubs.webpackUtils.getModuleResolutionPaths.returns( 'loader-resolution-paths' );
 
 		stubs.fs.readdirSync.withArgs( '/process/cwd/sample' ).returns( [
 			'ckeditor.js',
@@ -97,6 +100,27 @@ describe( 'lib/utils/get-webpack-config-server', () => {
 		const config = getWebpackConfigServer( { cwd } );
 
 		expect( config.resolve.extensions ).to.deep.equal( [ '.ts', '...' ] );
+	} );
+
+	it( 'resolves correct module paths', () => {
+		const config = getWebpackConfigServer( { cwd } );
+
+		expect( config.resolve ).to.have.property( 'modules', 'loader-resolution-paths' );
+	} );
+
+	it( 'resolves correct loader paths', () => {
+		const config = getWebpackConfigServer( { cwd } );
+
+		expect( config.resolveLoader ).to.have.property( 'modules', 'loader-resolution-paths' );
+	} );
+
+	it( 'calls "getModuleResolutionPaths" with correct arguments', () => {
+		getWebpackConfigServer( { cwd } );
+
+		expect( stubs.webpackUtils.getModuleResolutionPaths.callCount ).to.equal( 1 );
+		const normalizedArgument = stubs.webpackUtils.getModuleResolutionPaths.firstCall.firstArg
+			.split( path.sep ).join( path.posix.sep );
+		expect( normalizedArgument.endsWith( '/packages/ckeditor5-package-tools/lib/utils/../..' ) ).to.equal( true );
 	} );
 
 	it( 'processes the "ckeditor.js" file', () => {

--- a/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
@@ -231,9 +231,18 @@ describe( 'lib/utils/webpack-utils', () => {
 		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
 			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/../../node_modules' );
 
-			expect( stubs.path.resolve.callCount ).to.equal( 1 );
+			expect( stubs.path.resolve.callCount ).to.equal( 2 );
 			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
 				[ 'root/directory', '..', '..', 'node_modules' ]
+			);
+		} );
+
+		it( 'loads "node_modules" from root of the repository that uses the "ckeditor5-package-tools" package', () => {
+			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../../node_modules' );
+
+			expect( stubs.path.resolve.callCount ).to.equal( 2 );
+			expect( stubs.path.resolve.getCall( 1 ).args ).to.deep.equal(
+				[ 'root/directory', '..', '..', '..', 'node_modules' ]
 			);
 		} );
 	} );

--- a/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
@@ -228,30 +228,12 @@ describe( 'lib/utils/webpack-utils', () => {
 			expect( moduleResolutionPaths[ 0 ] ).to.equal( 'node_modules' );
 		} );
 
-		it( 'loads "node_modules" from root of the "ckeditor5-package-tools" package', () => {
-			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/node_modules' );
-
-			expect( stubs.path.resolve.callCount ).to.equal( 3 );
-			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
-				[ 'root/directory', 'node_modules' ]
-			);
-		} );
-
 		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
-			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../node_modules' );
+			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/../../node_modules' );
 
-			expect( stubs.path.resolve.callCount ).to.equal( 3 );
-			expect( stubs.path.resolve.getCall( 1 ).args ).to.deep.equal(
+			expect( stubs.path.resolve.callCount ).to.equal( 1 );
+			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
 				[ 'root/directory', '..', '..', 'node_modules' ]
-			);
-		} );
-
-		it( 'loads "node_modules" from root of the repository that uses the "ckeditor5-package-tools" package', () => {
-			expect( moduleResolutionPaths[ 3 ] ).to.equal( 'root/directory/../../../node_modules' );
-
-			expect( stubs.path.resolve.callCount ).to.equal( 3 );
-			expect( stubs.path.resolve.getCall( 2 ).args ).to.deep.equal(
-				[ 'root/directory', '..', '..', '..', 'node_modules' ]
 			);
 		} );
 	} );

--- a/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
@@ -231,18 +231,9 @@ describe( 'lib/utils/webpack-utils', () => {
 		it( 'loads "node_modules" from root of the "ckeditor5-package-tools" package', () => {
 			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/node_modules' );
 
-			expect( stubs.path.resolve.callCount ).to.equal( 2 );
+			expect( stubs.path.resolve.callCount ).to.equal( 1 );
 			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
 				[ 'root/directory', 'node_modules' ]
-			);
-		} );
-
-		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
-			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../node_modules' );
-
-			expect( stubs.path.resolve.callCount ).to.equal( 2 );
-			expect( stubs.path.resolve.getCall( 1 ).args ).to.deep.equal(
-				[ 'root/directory', '..', '..', 'node_modules' ]
 			);
 		} );
 	} );

--- a/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
@@ -228,21 +228,21 @@ describe( 'lib/utils/webpack-utils', () => {
 			expect( moduleResolutionPaths[ 0 ] ).to.equal( 'node_modules' );
 		} );
 
-		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
-			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/../../node_modules' );
+		it( 'loads "node_modules" from root of the "ckeditor5-package-tools" package', () => {
+			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/node_modules' );
 
 			expect( stubs.path.resolve.callCount ).to.equal( 2 );
 			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
-				[ 'root/directory', '..', '..', 'node_modules' ]
+				[ 'root/directory', 'node_modules' ]
 			);
 		} );
 
-		it( 'loads "node_modules" from root of the repository that uses the "ckeditor5-package-tools" package', () => {
-			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../../node_modules' );
+		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
+			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../node_modules' );
 
 			expect( stubs.path.resolve.callCount ).to.equal( 2 );
 			expect( stubs.path.resolve.getCall( 1 ).args ).to.deep.equal(
-				[ 'root/directory', '..', '..', '..', 'node_modules' ]
+				[ 'root/directory', '..', '..', 'node_modules' ]
 			);
 		} );
 	} );

--- a/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
+++ b/packages/ckeditor5-package-tools/tests/utils/webpack-utils.js
@@ -24,7 +24,8 @@ describe( 'lib/utils/webpack-utils', () => {
 
 		stubs = {
 			path: {
-				join: sinon.stub().callsFake( ( ...chunks ) => chunks.join( '/' ) )
+				join: sinon.stub().callsFake( ( ...chunks ) => chunks.join( '/' ) ),
+				resolve: sinon.stub().callsFake( ( ...chunks ) => chunks.join( '/' ) )
 			},
 			getThemePath: sinon.stub(),
 			devUtils: {
@@ -213,6 +214,45 @@ describe( 'lib/utils/webpack-utils', () => {
 				expect( '/Users/ckeditor/ckeditor5-foo/theme/icons/ckeditor.html' ).to.not.match( loader.test );
 				expect( 'C:\\Users\\ckeditor\\ckeditor5-foo\\theme\\icons\\ckeditor.html' ).to.not.match( loader.test );
 			} );
+		} );
+	} );
+
+	describe( 'getModuleResolutionPaths()', () => {
+		let moduleResolutionPaths;
+
+		beforeEach( () => {
+			moduleResolutionPaths = webpackUtils.getModuleResolutionPaths( 'root/directory' );
+		} );
+
+		it( 'loads "node_modules" directly', () => {
+			expect( moduleResolutionPaths[ 0 ] ).to.equal( 'node_modules' );
+		} );
+
+		it( 'loads "node_modules" from root of the "ckeditor5-package-tools" package', () => {
+			expect( moduleResolutionPaths[ 1 ] ).to.equal( 'root/directory/node_modules' );
+
+			expect( stubs.path.resolve.callCount ).to.equal( 3 );
+			expect( stubs.path.resolve.getCall( 0 ).args ).to.deep.equal(
+				[ 'root/directory', 'node_modules' ]
+			);
+		} );
+
+		it( 'loads "node_modules" from root of the "ckeditor5-package-generator" repository', () => {
+			expect( moduleResolutionPaths[ 2 ] ).to.equal( 'root/directory/../../node_modules' );
+
+			expect( stubs.path.resolve.callCount ).to.equal( 3 );
+			expect( stubs.path.resolve.getCall( 1 ).args ).to.deep.equal(
+				[ 'root/directory', '..', '..', 'node_modules' ]
+			);
+		} );
+
+		it( 'loads "node_modules" from root of the repository that uses the "ckeditor5-package-tools" package', () => {
+			expect( moduleResolutionPaths[ 3 ] ).to.equal( 'root/directory/../../../node_modules' );
+
+			expect( stubs.path.resolve.callCount ).to.equal( 3 );
+			expect( stubs.path.resolve.getCall( 2 ).args ).to.deep.equal(
+				[ 'root/directory', '..', '..', '..', 'node_modules' ]
+			);
 		} );
 	} );
 } );

--- a/scripts/ci/travis-check.js
+++ b/scripts/ci/travis-check.js
@@ -70,7 +70,9 @@ start();
  */
 async function start() {
 	await testBuild( 'js', 'npm' );
-	await testBuild( 'ts', 'yarn', 'CustomPluginName400' );
+	await testBuild( 'ts', 'npm', 'CustomPluginName' );
+	await testBuild( 'js', 'yarn', 'customPluginName400' );
+	await testBuild( 'ts', 'yarn' );
 
 	if ( foundError ) {
 		console.log( '\n' + chalk.red( 'Found errors during the verification. Please, review the log above.' ) );

--- a/scripts/ci/travis-check.js
+++ b/scripts/ci/travis-check.js
@@ -95,20 +95,12 @@ async function testBuild( lang, packageManager, customPluginName ) {
 		'--dev', '--verbose', '--lang', lang, `--use-${ packageManager }`
 	];
 
+	const expectedPublishFiles = getExpectedFiles( EXPECTED_PUBLISH_FILES, lang, customPluginName );
+	const expectedSrcDirFiles = getExpectedFiles( EXPECTED_SRC_DIR_FILES, lang, customPluginName );
+
 	if ( customPluginName ) {
 		testSetupInfoMessage += ` with custom plugin name: [${ customPluginName }]`;
 		packageBuildCommand.push( '--plugin-name', customPluginName );
-
-		const filenameArrays = [
-			...Object.values( EXPECTED_PUBLISH_FILES ),
-			...Object.values( EXPECTED_SRC_DIR_FILES )
-		];
-
-		filenameArrays.forEach( filenameArray => {
-			filenameArray.forEach( ( filename, index, arr ) => {
-				arr[ index ] = filename.replace( 'testpackage', customPluginName.toLowerCase() );
-			} );
-		} );
 	}
 
 	logProcess( testSetupInfoMessage + '.' );
@@ -129,10 +121,10 @@ async function testBuild( lang, packageManager, customPluginName ) {
 	logProcess( 'Verifying release process...' );
 	const { stderr } = executeCommand( [ 'npm', 'publish', '--dry-run' ], { cwd: NEW_PACKAGE_DIRECTORY, pipeStderr: true } );
 	console.log( stderr );
-	checkFileList( stderr, lang );
+	checkFileList( stderr, expectedPublishFiles );
 
 	logProcess( 'Verifying post release cleanup...' );
-	verifyPublishCleanup( lang );
+	verifyPublishCleanup( lang, expectedSrcDirFiles );
 
 	logProcess( 'Starting the development servers and verifying the sample builds...' );
 	await Promise.all( [
@@ -306,9 +298,9 @@ function logProcess( message ) {
  * Checks whether output of "npm publish" contains correct files.
  *
  * @param {String} output
- * @param {string} lang
+ * @param {Object} expectedPublishFiles
  */
-function checkFileList( output, lang ) {
+function checkFileList( output, expectedPublishFiles ) {
 	const match = output.match( /Tarball Contents.+\n(?<lines>[\s\S]+)\n.+Tarball Details/ );
 
 	if ( !match ) {
@@ -321,8 +313,8 @@ function checkFileList( output, lang ) {
 
 	const files = match.groups.lines.split( '\n' ).map( string => string.trim().split( ' ' ).pop() );
 
-	const missingFiles	= EXPECTED_PUBLISH_FILES[ lang ].filter( item => !files.includes( item ) );
-	const excessFiles	= files.filter( item => !EXPECTED_PUBLISH_FILES[ lang ].includes( item ) );
+	const missingFiles	= expectedPublishFiles.filter( item => !files.includes( item ) );
+	const excessFiles	= files.filter( item => !expectedPublishFiles.includes( item ) );
 
 	if ( !missingFiles.length && !excessFiles.length ) {
 		console.log( chalk.green( 'Files staged for publishing verified successfully.' ) );
@@ -350,8 +342,9 @@ function checkFileList( output, lang ) {
  * - There should be no leftover build files in "src" directory.
  *
  * @param {String} lang
+ * @param {Object} expectedSrcDirFiles
  */
-function verifyPublishCleanup( lang ) {
+function verifyPublishCleanup( lang, expectedSrcDirFiles ) {
 	// "package.json" check.
 	const pkgJsonPath = path.join( NEW_PACKAGE_DIRECTORY, 'package.json' );
 	const pkgJsonRaw = fs.readFileSync( pkgJsonPath, 'utf-8' );
@@ -369,7 +362,7 @@ function verifyPublishCleanup( lang ) {
 	// "src" directory check.
 	const srcDirPath = path.join( NEW_PACKAGE_DIRECTORY, 'src' );
 	const excessFiles = fs.readdirSync( srcDirPath )
-		.filter( file => !EXPECTED_SRC_DIR_FILES[ lang ].includes( file ) );
+		.filter( file => !expectedSrcDirFiles.includes( file ) );
 
 	if ( excessFiles.length ) {
 		console.log( chalk.red( 'Excess files after publishing in "src" directory:' ) );
@@ -381,4 +374,21 @@ function verifyPublishCleanup( lang ) {
 	if ( hasCorrectEntryPoint && !excessFiles.length ) {
 		console.log( chalk.green( 'Post release cleanup successful.' ) );
 	}
+}
+
+/**
+ * Returns an array of expected files for a given language.
+ * Custom plugin name is also accounted for, if its specified.
+ *
+ * @param {Object} expectedFilesObject
+ * @param {String} lang
+ * @param {String|undefined} customPluginName
+ * @returns {Array<String>}
+ */
+function getExpectedFiles( expectedFilesObject, lang, customPluginName ) {
+	if ( !customPluginName ) {
+		return expectedFilesObject[ lang ];
+	}
+
+	return expectedFilesObject[ lang ].map( filename => filename.replace( 'testpackage', customPluginName.toLowerCase() ) );
 }

